sources/packages/backend/apps/api/src/route-controllers/application-restriction-bypass/application-restriction-bypass.aest.controller.ts

```diff
@@ -50,4 +90,171 @@ export class ApplicationRestrictionBypassAESTController extends BaseController {
+  }
+
+  /**
+   * Gets available student restrictions to bypass for a given application.
+   * @param applicationId id of the application to retrieve restriction bypasses.
+   * @returns application restriction bypasses.
+   */
+  @Get("application/:applicationId/options-list")
+  async getAvailableStudentRestrictionsToBypass(
+    @Param("applicationId", ParseIntPipe) applicationId: number,
+  ): Promise<AvailableStudentRestrictionAPIOutDTO> {
+    const availableRestrictionsToBypass =
+      await this.applicationRestrictionBypassService.getAvailableStudentRestrictionsToBypass(
+        applicationId,
+      );
+    return {
+      availableRestrictionsToBypass: availableRestrictionsToBypass.map(
+        (item: StudentRestriction) => ({
+          studentRestrictionId: item.id,
+          restrictionCode: item.restriction.restrictionCode,
+          studentRestrictionCreatedAt: item.createdAt,
+        }),
+      ),
+    };
+  }
+
+  /**
+   * Creates a new application restriction bypass.
+   * @param payload payload of the application restriction bypass.
+   * @param userToken user token.
+   * @returns application restriction bypass.
+   */
+  @ApiUnprocessableEntityResponse({
+    description:
+      "Cannot create a bypass when there is an active bypass for the same active student restriction id or " +
+      "could not find student restriction for the given id or " +
+      "cannot create a bypass when student restriction is not active  or " +
+      "cannot create a bypass when application is in invalid state.",
+  })
+  @Roles(Role.AESTBypassStudentRestriction)
+  @Post()
+  async bypassRestriction(
+    @Body() payload: BypassRestrictionAPIInDTO,
+    @UserToken() userToken: IUserToken,
+  ): Promise<PrimaryIdentifierAPIOutDTO> {
+    try {
+      const applicationRestrictionBypass =
+        await this.applicationRestrictionBypassService.bypassRestriction(
+          payload,
+          userToken.userId,
+        );
+      return { id: applicationRestrictionBypass.id };
+    } catch (error) {
+      if (error instanceof CustomNamedError) {
+        switch (error.name) {
+          case ACTIVE_BYPASS_FOR_STUDENT_RESTRICTION_ALREADY_EXISTS:
+            throw new UnprocessableEntityException(
+              "Cannot create a bypass when there is an active bypass for the same active student restriction.",
+            );
+          case STUDENT_RESTRICTION_NOT_FOUND:
+            throw new UnprocessableEntityException(
+              "Could not find student restriction for the given id.",
+            );
+          case STUDENT_RESTRICTION_IS_NOT_ACTIVE:
+            throw new UnprocessableEntityException(
+              "Cannot create a bypass when student restriction is not active.",
+            );
+          case APPLICATION_IN_INVALID_STATE_FOR_APPLICATION_RESTRICTION_BYPASS_CREATION:
+            throw new UnprocessableEntityException(
+              "Cannot create a bypass when application is in invalid state.",
+            );
```

Comment:
The usual approach for the error names would be creating the necessary ones to make decisions.
If the only condition applied to different error names is the message, the usual approach is the below.
https://github.com/bcgov/SIMS/blob/7c7f20cf45757b5f6e3021b4e60cf2e184b815e1/sources/packages/backend/apps/api/src/route-controllers/application-offering-change-request/application-offering-change-request.institutions.controller.ts#L296-L312

Here, the ones that are expected to be handled by the consumer are thrown using the below.
```ts
new ApiProcessError(error.message, error.name)
```
Ones that are not expected to be processed by the consumer are thrown using the below.
```ts
throw new NotFoundException(error.message);
```

Reply:
Thanks for doing the change. For my understanding all the errors are now using `ApiProcessError` because there is an expectation to that they will need to be handled by the consumer/UI (showing something for the user)?

Reply:
We have expected "business" errors and some that are not.
For instance, a NOT FOUND usually is not returned using the `ApiProcessError`.
In essence we can return every single API error in a `ApiProcessError` but we are not following this.
The rule is: expected messages that are required to be shown to the user are wrapped in a `ApiProcessError`, every other will be just a regular exception.

Reply:
It will not be a blocker, but the understanding should be clear.

Reply:
IMO all this messages are meaningful for the user if that happens. I know it's unlikely to happen but at least when they happen the user might have a clue of the issue or tell it to a systems administrator. Otherwise, what do you suggest?

Reply:
> We have expected "business" errors and some that are not. For instance, a NOT FOUND usually is not returned using the `ApiProcessError`. In essence we can return every single API error in a `ApiProcessError` but we are not following this. The rule is: expected messages that are required to be shown to the user are wrapped in a `ApiProcessError`, every other will be just a regular exception.
> 
> For instance, are the below ones meaningful from user perspective?
> 
> * STUDENT_RESTRICTION_NOT_FOUND
> * APPLICATION_IN_INVALID_STATE_FOR_APPLICATION_RESTRICTION_BYPASS_CREATION

My suggestion is explained above and it represents what I believe the solution has been following.
From a point of view, as I mentioned earlier, we can return every single API error as an `ApiProcessError`. The system administrator should and will not rely on what the user sees. The application logs must provide the complete error and should be the source of thurh for investigation.

Please keep all of these as they are. I am closing the conversation.